src/socket.rs

```diff
@@ -242,14 +253,36 @@ where
         }
     }
 
+    async fn add_outbound_connection(&self, cid: &ConnectionId<P>, events_tx: mpsc::UnboundedSender<StreamEvent>) {
+        loop {
+            let num_connections = self.conns.read().unwrap().len();
+            if num_connections < self.max_conns {
+                break;
+            }
+            self.outbound_dropped.notified().await;
```

Comment:
Is this holding the read() lock across the await? Do we need something like:

```suggestion
            let num_connections = {
                self.conns.read().unwrap().len()
            };
            if num_connections < self.max_conns {
                break;
            }
            self.outbound_dropped.notified().await;
```

It hasn't seemed to cause problems during testing, but it's obviously worth double-checking.